core/src/main/scala/kafka/coordinator/group/MemberMetadata.scala

```diff
@@ -78,6 +79,13 @@ private[group] class MemberMetadata(val memberId: String,
     }
   }
 
+  def shouldKeepAlive(deadlineMs: Long): Boolean = {
+    if (awaitingJoinCallback != null)
```

Comment:
Hmm.. in `GroupCoordinator` code after we've triggered the callback, we actually did not set it to null. Maybe this does not affect the correctness of the logic but I'm a bit concerned it is vulnerable to bugs in the future, maybe we can just remove the callback after triggers it above (see my other comment)?